### ISSUE TITLE
WebGL potential support

### ISSUE THREAD

**Issue by pmp-p** [szabolcsdombi/zengl#34]
Hi, would it be possible to have/keep a branch alive with version 300 es ?
it seems the 1.10.2 release more close to it

**Comment by szabolcsdombi** [comment]
Hi, indeed 2.x dropped support for OpenGL 3.3 and 1.10 still has it but I think 2.x would be a better candidate for WebGL support as it runs perfectly on ANGLE and was tested with OpenGL ES 3.0 and 3.1.
In case something is missing I can add some workaround for it.

In the C++ code [here](https://github.com/szabolcsdombi/ModernGL-ZenGL-with-GLES) I implement the missing GL functions, those are very simple ones part of OpenGL 4.5 but not OpenGL 3.3 despite all the functionality is available.
This angle support soon will be part of glcontext or just ZenGL itself.

The major difference between zengl 2.x and zengl 1.x is the extra support for storage buffers, compute shaders, and the use of explicit resource binding.

For WebGL the extra part may not work but the core of ZenGL was not changed much (except that I use the latest OpenGL calls, but those can be implemented by the loader using the older methods).

Do you have some build pipeline already prepared to such port? I would be interested in helping the build pass for ZenGL.

**Comment by szabolcsdombi** [comment]
- for explicit binding this might mean an optional `layout={}` parameter should be restored from 1.10.0

**Comment by pmp-p** [comment]
The build passes ( with https://github.com/pygame-web/python-wasm-sdk  and running `/opt/python-wasm-sdk/python3-wasm setup.py bdist_wheel` ) for glcontext / zengl
but there's still quite a lot of mismatched functions pointer signatures : wasm/webgl hates that and it is only detectable only at runtime, unless using a C++ parser to check source code.

**Comment by szabolcsdombi** [comment]
On ubuntu 22.04 with Python 3.11 breaks at `setup.py config cython` for pygame.

```Dockerfile
FROM ubuntu:22.04
RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
        build-essential gdb lcov pkg-config \
        libbz2-dev libffi-dev libgdbm-dev libgdbm-compat-dev liblzma-dev \
        libncurses5-dev libreadline6-dev libsqlite3-dev libssl-dev \
        lzma lzma-dev tk-dev uuid-dev zlib1g-dev \
        git bash curl lz4 pv
RUN git clone https://github.com/python/cpython -b 3.11 && cd /cpython && ./configure --enable-optimizations && make -j4
RUN /cpython/python -m ensurepip && /cpython/python -m pip install git+https://github.com/cython/cython.git
RUN git clone https://github.com/pygame/pygame
RUN curl -sL https://github.com/pygame-web/python-wasm-sdk/releases/download/3.1.33.0/python3.11-wasm-sdk-Ubuntu-22.04.tar.lz4 -o sdk.tar.lz4
RUN tar xfvP sdk.tar.lz4 --use-compress-program=lz4
WORKDIR /pygame
RUN /cpython/python setup.py config cython
```

**Comment by pmp-p** [comment]
ho sorry the readme is very obsolete, python-wasm-sdk prebuilt release is meant to be used directly in CI eg from https://github.com/pygame-community/pygame-ce/blob/main/.github/workflows/build-emsdk.yml

if you want to rebuild the sdk locally ( which i do as i don't use ubuntu locally ) just do `EMFLAVOUR=latest ./python-wasm-sdk.sh`, can take up to 1 hour

**Comment by szabolcsdombi** [comment]
I see, I just have to use the /opt/ files only.

```Dockerfile
FROM ubuntu:22.04
RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y curl git lz4
RUN curl -sL https://github.com/pygame-web/python-wasm-sdk/releases/download/3.1.33.0/python3.11-wasm-sdk-Ubuntu-22.04.tar.lz4 -o sdk.tar.lz4 &&\
    tar xfP sdk.tar.lz4 --use-compress-program=lz4 &&\
    rm sdk.tar.lz4
RUN git clone https://github.com/pygame/pygame /pygame
WORKDIR /pygame
RUN /opt/python-wasm-sdk/python3-wasm setup.py build
# apt-get install libsdl2-dev libsdl-image1.2-dev libfreetype6-dev
```

This still fails, installing freetype as last did not even help.

**Comment by szabolcsdombi** [comment]
I just need a dockerfile that can finally run as the webserver having zengl built anywhere in the build steps so I can debug it myself.

**Comment by pmp-p** [comment]
CI passed on zengl with https://github.com/pmp-p/zengl-wasm/blob/main/.github/workflows/emsdk.yml, 2.x seems less compatible with webgl than 1.10.2 but not GLE/GLES really not my expertise.

**Comment by pmp-p** [comment]
got my triangle ! https://pygame-web.github.io/showroom/pypad_git.html?-#src/test_zengl.py  ( use `rx out.png` to get the framebuffer )

maybe for the glcontext loader replacement ideally it should be static linking see
https://emscripten.org/docs/optimizing/Optimizing-WebGL.html#optimizing-load-times-and-other-best-practices.
runtime linking is problematic some functions as missing in GLES3

**Comment by szabolcsdombi** [comment]
static linking and replacement functions for the used but missing ones is in the `glcontext_emscriptem` I look into the CI you linked and put it together for the static linked gl functions.

runtime loading the functions should not affect the load time much, there are 121 functios to be looked up by name,  `glcontext_emscriptem` does it with O(N^2) hopefully at the scale of microseconds.

🎉 for the hello triangle

**Comment by szabolcsdombi** [comment]
Good News!

Got the complete feature set of ZenGL working without emscripten glue code for webgl2.
We can support any number of canvases, and it does not require pyodide to be extended in any way.

Some parts are not yet implemented, but I can clear them up when porting the whole example library.

**Comment by szabolcsdombi** [comment]
![image](https://user-images.githubusercontent.com/11232402/224429947-f9062f2c-5ab7-4d67-9680-eab315cbbf21.png)

**Comment by szabolcsdombi** [comment]
As it turns out Pyodide has an `import js` that can access the window globas. At the moment I pass the WebGL2RenderingContext in it. This could be improved as this variable will be baked in the `zengl.context()` anyways.
I found no other way to pass it as a parameter to my DefaultLoader.

The code built for the web is patched in the Dockerfile. The stubs are removed and the webgl loader is concatenated to the _zengl helper module.

To have multi canvas support the window.gl must be different for each `zengl.context()` call. Previously created contexes keep working as expected.

**Comment by pmp-p** [comment]
indeed just need to pass the GL context value, on [android](https://raw.githubusercontent.com/pmp-p/pydk/master/docs/Panda3D/pandapk-step2.png) i do that with an env var comparable to SDL_WINDOWID.

( i think python packages should not use pyodide code to do low level stuff. There are a lot of other python implementations around )

what about just zengl.context(context:int) and a null loader ?
I guess it's possible to read context configuration with GL commands

**Comment by szabolcsdombi** [comment]
> what about just zengl.context(context:int) and a null loader ?

Here is the signature of the context method:

```py
class ContextLoader(Protocol):
    def load_opengl_function(name: str) -> int: ...

def context(loader: ContextLoader | None = None) -> Context: ...
def loader(headless: bool = False) -> ContextLoader: ...
```

`zengl.context()` is equivalent to `zengl.context(zengl.loader())` that is equivalent to `zengl.context(_zengl.DefaultBackend())` that can be easily replaced by a custom build just like i did it in:
https://github.com/szabolcsdombi/zengl/blob/30422dde9c860bc9babcad1013ad44b020e441f4/examples/webgl/Dockerfile#L9

For Pyodide the webgl.py (not part of zengl itself) is appended to the helper module with `cat webgl.py >> _zengl.py ` hence `zengl.context()` defaults to using it.

Explicit is often better so it is also possible to just call:

```py
ctx = zengl.context(MyCustomLoader())
```

My intentions with the append was to get the example code working as is without a change. I do not plan to include any complex loaders in the main source code. For that glcontext or any other library is a better fit.

Thank you for the information provided above, it helped me with getting zengl working on the web.

I am interested in testing out zengl with your pygame build, but it is still broken on my side in my dockerfile.
For now at least I can test the full feature set without being blocked on this one.
It would be nice to have pygame+zengl code working as is on the web.

**Comment by pmp-p** [comment]
> It would be nice to have pygame+zengl code working as is on the web.

For that it's quite simple, don't use pyodide idiomatics or javascript but emscripten EGL+offscreen instead.

Though publishing your wheel artefacts in the CI would surely help me to bridge the gap beetween pyodide and python-wasm (pygbag), as pygbag "platform" module and pyodide "js" are not that different.


i'm ready to help but building pyodide is time consumming while a pygbag build take less than 30 seconds and does not require docker.

**Comment by szabolcsdombi** [comment]
I need the following to be able to work efficiently on pygbag examples:

A Dockerfile which can compile zengl. This is partially done with the help of your toolbox:

```Dockerfile
FROM ubuntu:22.04
RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y curl git lz4
RUN curl -sL https://github.com/pygame-web/python-wasm-sdk/releases/download/3.1.33.0/python3.11-wasm-sdk-Ubuntu-22.04.tar.lz4 -o sdk.tar.lz4 &&\
    tar xfP sdk.tar.lz4 --use-compress-program=lz4 &&\
    rm sdk.tar.lz4
WORKDIR /zengl/
COPY . .
RUN /opt/python-wasm-sdk/python3-wasm setup.py build
```

I mentiened eralier I cannot build pygame with it, but you just linked the archives for pygame_static so I no longer have to build pygame. I can use the prebuilt one.
So I can just grab the .whl it with `wget`.

```Dockerfile
RUN wget https://github.com/pygame-web/archives/blob/main/repo/pkg/pygame_static-1.0-cp311-cp311-wasm32_mvp_emscripten.whl
```

I can add the index.html, index.js, hello.py and serve it

```Dockerfile
COPY some/path/index.html some/path/index.js some/path/hello.py /zengl/
CMD python -m http.server --bind 0.0.0.0 8000
```

Can you point me to an example where I get a minimal code for the index.js and hello.py?

```sh
docker build . -t temp && docker run --rm -it -p 8000:8000 temp
```

Please note that I use Docker on a remote computer.
I am only using COPY commands to get data in the docker image.
No volume mounts or similar.
I can access the containers port with my browser.

I almost never install any software on computer directly.
All the tools I need can run with docker remotely.

**Comment by pmp-p** [comment]
> Can you point me to an example where I get a minimal code for the index.js and hello.py?

pygbag build index.html from a folder/main.py + a template, you can grab index.html in folder/build/web
 or you can write your own .py as .html eg https://github.com/pygame-web/showroom/tree/main/pygame-scripts  ( you can use "black -x" to format and "python -x" to run those directly )

The javascript files are quite static and are all hosted here https://github.com/pygame-web/archives/tree/main/0.7 or here for the git version https://github.com/pygame-web/pygbag/tree/gh-pages/0.0

Unlike pyodide typescript startup, pygbag's one is written in python and can be found in pythonrc.py

The equivalent of "index.js" is named "python.js"
